### PR TITLE
Packwerk checks for stale violations

### PR DESCRIPTION
**Pull request by cindygshopify**
## What are you trying to accomplish?
Adding feature for Packwerk checking stale violations. This should be incorporated into CI in the future to make sure that PRs don't pass CI if there are stale violations. Linked issue here: https://github.com/Shopify/packwerk/issues/12

## What approach did you choose and why?

We had to refactor the code a bit to create this feature. We needed a lot of functionality in the UpdatingDeprecatedReferences.rb class but putting the change in that class didn't seem right. So CacheDeprecatingReferences.rb was created as a parent class that both UpdatingDeprecatedReferences.rb and DetectStaleDeprecatedReferences.rb now inherit from. Functionality common to both now belongs in CacheDeprecatingReferences.rb and functionality specific to updating vs detecting stale violations are now in their respective classes.

When executing the command, Packwerk will go through all the loaded files and check for violations in the files which it then stores in DetectStaleDeprecatedReferences.rb. At the end when all files were processed, we compare the deprecated_references.yml file of each component loaded to the actual violations generated. If there are references in deprecated_references.yml that aren't in the generated violations, then there are stale violations. The program will exits with status code 1. If there are no stale violations, program exits with status code 0.

## What should reviewers focus on?

Any issues with the code refactor? Are there stale violation cases we may have missed with this code change?

## Tophatting

My tophatting instructions use Packwerk in the shopify/shopify repo, but you can use another repo if you prefer

1) Pull this branch into local
2) Change the gemfile of local shopify/shopify repo to point to this branch
3) Modify dev.yml of shopify/shopify to include the detect-stale violations command like so:
Screen Shot 2020-11-03 at 5 46 32 PM
4) Generate a stale violation in shopify/shopify. 

- You can do this by changing package.yml of a component, enforcing dependencies on it and run Packwerk update to generate a deprecated_references.yml for that component which has boundary violations. 

- After you have a list of dependency violations for that component, you can add the component which the violation is coming from as a dependency in the  package.yml of your component

5) Run `dev packages detect-stale-violations path\to\your\component`
6) Check that it outputs stale violations were made
7) Run `$?` to check status of last command (0 if no stale violations, 1 if there were stale violations)

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code refactor (non-breaking change that doesn't add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Additional Release Notes

Include any notes here to include in the release description. For example, if you selected "breaking change" above, leave notes on how users can transition to this version.

If no additional notes are necessary, delete this section or leave it unchanged.

## Checklist

- [X] It is safe to simply rollback this change.

### CODE REVIEW

File: lib/packwerk/deprecated_references.rb
Comment (tomstuart) [inline lib/packwerk/deprecated_references.rb:51]: This whole structure…

```ruby
deprecated_references.each_key do |package|
  deprecated_references[package].each_key do |constant_name|
    # …
    stale_violations = … deprecated_references.dig(package, constant_name, "files") …
    # …
    if stale_violations.present?
      return true
    end
  end
end
false
```

…could instead be written as:

```ruby
deprecated_references.any? do |package, package_violations|
  package_violations.any? do |constant_name, entries_for_file|
    # …
    stale_violations = … entries_for_file["files"] …
    # …
    stale_violations.present?
  end
end
```

The changes are:

1. use `Enumerable#any?` to check as many elements as necessary and decide whether to return `true` or `false` overall, rather than using `Hash#each_key` and breaking out of the loop explicitly ourselves;
2. use the values as well as the keys from the outer `deprecated_references` hash as we iterate over it, rather than throwing away the values and then using the keys to look them up again;
3. likewise, use the values as well as the keys from the inner `package_violations` hash as we iterate over it rather than `dig`ing into the outer `deprecated_references`; and
4. implicitly returning `stale_violations.present?` from the inner block (for use by `#any?`) rather than breaking out with an explicit `return`.

I think the second version makes it easier to see at a glance that we’re iterating over every violation in `deprecated_references` (two levels deep) and checking whether `any?` of them are stale. It also happens to be a little bit more concise which is a nice bonus. What do you think?

File: lib/packwerk/detect_stale_deprecated_references.rb
Comment (tomstuart) [inline lib/packwerk/detect_stale_deprecated_references.rb:9]: You could do…

```suggestion
      @deprecated_references.values.any? do |deprecated_references_file|
```

…instead if you prefer that.

File: lib/packwerk/updating_deprecated_references.rb
Comment (tomstuart) [inline lib/packwerk/updating_deprecated_references.rb:31]: 👀

File: test/unit/deprecated_references_test.rb
Comment (tomstuart) [inline test/unit/deprecated_references_test.rb:28]: These four tests are (very nicely! 👏 ) unit testing just `DeprecatedReferences#stale_violations?`, which doesn’t depend on any of the other machinery being added simultaneously in `DetectStaleDeprecatedReferences` and `Cli`. Would you consider pulling this out into a separate commit, i.e. first implement & test `DeprecatedReferences#stale_violations?`, then add everything else afterwards? Having these in separate commits would make it more obvious which changes go together.

File: test/unit/detect_stale_deprecated_references_test.rb
Comment (tomstuart) [inline test/unit/detect_stale_deprecated_references_test.rb:9]: Ditto my comment about the other unit tests: after `DeprecatedReferences#stale_violations?` is added, these tests could go in their own commit along with the `DetectStaleDeprecatedReferences#stale_violations?` implementation they’re testing, and then a third commit could finally implement `Cli#detect_stale_violations`. As well as making it clearer which changes go together, this sequence of commits would make it clearer which pieces build on top of which other pieces.

File: lib/packwerk/cli.rb
Comment (tomstuart) [inline lib/packwerk/cli.rb:183]: There’s no test for this. There are already other subcommands (e.g. `packwerk check`) which have no unit test, but those are covered by the integration tests. Is there an easy way to have _some_ automated test coverage of `packwerk detect-stale-violations`, whether unit or integration, so that it doesn’t get accidentally broken?

**Review by tomstuart — COMMENTED** [review]
Thank you for doing this, Cindy!

No showstoppers here but I had a few ideas while reading through your code so I’ve added some comments for now in case you’d like to discuss them. Let me know what you think.

File: lib/packwerk/cache_deprecated_references.rb
Comment (doodzik) [inline lib/packwerk/cache_deprecated_references.rb:1]: I think it would be good that new files to this codebase are in `strict` mode. A lot of the benefits from sorbet only arise if we have enough type information available.

File: lib/packwerk/cli.rb
Comment (doodzik) [reply lib/packwerk/cli.rb:183]: @rochlefebvre was working on an improvement in this PR https://github.com/Shopify/packwerk/pull/59 so it could make sense to have a look at that.

We are adding more code to the cli class, which makes it harder to maintain. I think it would make sense to move the logic of this command into a separate class, which would make it easier to reason about the code and subsequently test.

File: lib/packwerk/detect_stale_deprecated_references.rb
Comment (doodzik) [inline lib/packwerk/detect_stale_deprecated_references.rb:8]: Same as my previous comment. Can you add type annotations to this class?

File: lib/packwerk/cache_deprecated_references.rb
Comment (cindygshopify) [reply lib/packwerk/cache_deprecated_references.rb:1]: While that would be nice, these methods are extracted from what used to be UpdatingDeprecatedReferences class and unfortunately I don't have enough context to know what the signatures for these classes would be.

File: lib/packwerk/detect_stale_deprecated_references.rb
Comment (cindygshopify) [reply lib/packwerk/detect_stale_deprecated_references.rb:10]: done

File: lib/packwerk/cli.rb
Comment (cindygshopify) [reply lib/packwerk/cli.rb]: @doodzik I'm not against moving the logic of it into another class, but I'm not really sure what the class would be. Currently I just wrote an additional test for cli_test.rb as it seemed like that was the best place to put a test for this function as many of the other command line related tests are also there. Anyway, let me know if you think there should be changes. I considered it but also didn't want to break the current pattern.

File: test/unit/detect_stale_deprecated_references_test.rb
Comment (cindygshopify) [reply test/unit/detect_stale_deprecated_references_test.rb:9]: I went ahead and did this. let me know if you have any concerns

File: lib/packwerk/cli.rb
Comment (wildmaples) [inline lib/packwerk/cli.rb:73]: Why was this removed?

File: lib/packwerk/deprecated_references.rb
Comment (wildmaples) [inline lib/packwerk/deprecated_references.rb:10]: Probably didn't mean to include this line

File: lib/packwerk/cli.rb
Comment (cindygshopify) [reply lib/packwerk/cli.rb:73]: Testing purposes. The private keyword is added back later on in the file for methods that don't get called from `execute_command`

**Review by wildmaples — COMMENTED** [review]
Yay this is great work!

File: test/unit/cache_deprecated_references_test.rb
Comment (cindygshopify) [reply test/unit/cache_deprecated_references_test.rb:7]: well I simply since we extracted CacheDeprecatedReferences from UpdatingDeprecatedReferences, the tests that were in UpdatingDeprecatedReferences were extracted to CacheDeprecatedReferences as the methods that are being tested now belong to CacheDeprecatedReferences. Now the UpdatingDeprecatedReferencesclass  just has one method that is specific to its' own class, but because no test was ever written for that method, UpdatingDeprecatedReferences now doesn't have a test class.

File: lib/packwerk/cache_deprecated_references.rb
Comment (doodzik) [reply lib/packwerk/cache_deprecated_references.rb:1]: I think it is important that we continue adopting sorbet in this codebase because it will improve the maintainability of the code over time. It is a bit inconvenient to look up the signatures but it is beneficial in the long term.

You can add type signatures to this class pretty easily and change to strict:

```ruby
   
    sig { params(root_path: String, deprecated_references: T::Hash[String, Packwerk::DeprecatedReferences]) }
    def initialize(root_path, deprecated_references = {})
      @root_path = root_path
      @deprecated_references = T.let(deprecated_references, T::Hash[String, Packwerk::DeprecatedReferences])
    end

    sig { params(package: Package).returns(Packwerk::DeprecatedReferences) } 
    def deprecated_references_for(package)
      @deprecated_references[package] ||= Packwerk::DeprecatedReferences.new(
        package,
        deprecated_references_file_for(package),
      )
    end

    sig { params(package: Package).returns(String) } 
    def deprecated_references_file_for(package)
      File.join(@root_path, package.name, "deprecated_references.yml")
    end
```

File: lib/packwerk/detect_stale_deprecated_references.rb
Comment (doodzik) [reply lib/packwerk/detect_stale_deprecated_references.rb:10]: 🙌

File: lib/packwerk/cli.rb
Comment (doodzik) [inline lib/packwerk/cli.rb:185]: This class is getting too big and complicated to maintain. I think it would be good to apply the [scout rule](https://www.oreilly.com/library/view/97-things-every/9780596809515/ch08.html) here, leaving a class cleaner than you found it. 

I would suggest moving your changes to a separate class so we can stop the rot. This should also be pretty straightforward since it doesn't require a big refactor.

File: lib/packwerk/cli.rb
Comment (cindygshopify) [reply lib/packwerk/cli.rb:186]: I don't think that link is working for me. Which line were you referring to?

File: lib/packwerk/cli.rb
Comment (cindygshopify) [reply lib/packwerk/cli.rb:204]: Sorry, what do you mean by the Packwerk update command? Did you just want backticks for the output message about stale violations?

File: lib/packwerk/cli.rb
Comment (exterm) [reply lib/packwerk/cli.rb:185]: I think it makes sense for the `CLI` class to have one method per packwerk subcommand, however we can consider reducing the size of this method considerably by moving most of its implementation somewhere else.

IMO, `CLI` should just be a thin wrapper that does the CLI specific stuff (like managing the output stream and parsing parameters) and delegates to other classes for everything else.

File: lib/packwerk/cli.rb
Comment (exterm) [reply lib/packwerk/cli.rb:204]: I think this is what Maple means:

```suggestion
      msg = status ? "No stale violations detected" : "There were stale violations found, please run `packwerk update`"
```

File: lib/packwerk/cache_deprecated_references.rb
Comment (exterm) [inline lib/packwerk/cache_deprecated_references.rb:14]: I think we can actually make this class abstract - it should never be instantiated, just its child classes, right?

We could to that by using sorbet's `abstract!`, which [can be used to prevent a class from being instantiated.](https://sorbet.org/docs/abstract).

File: lib/packwerk/cli.rb
Comment (tomstuart) [reply lib/packwerk/cli.rb:185]: I agree that `CLI` is a mess and should be reorganised but am more ambivalent about whether it necessarily must happen in this PR. Could one of us open a new PR to do the cleanup, either before or after this one lands?

Comment (exterm) [reply lib/packwerk/cli.rb:185]: I agree with Tom - Actually that was what I had in mind but failed to address the point. I meant to say "what should happen to CLI is a larger refactor and should thus be done separately"

**Comment by cindygshopify** [comment]
I've pushed one more commit that moves the detect_stale_violations function that was in cli.rb into it's own class so we can get some separation of concerns in the Packwerk codebase. This required some refactoring of cli.rb as well by putting some common functions into a module. I've also updated the code to reflect some of the above mentioned comments, so please take a look. In the interest of time, I'd prefer not to drag out this PR more if it is related to further refactoring. I think this is better served by creating new GitHub issues around that so we can avoid scope creep.

File: lib/packwerk/cache_deprecated_references.rb
Comment (cindygshopify) [reply lib/packwerk/cache_deprecated_references.rb:1]: I attempted to do this as it seemed like it would be low hanging fruit we could ship with this PR, but these signatures generated some really weird typecheck errors that should be further investigated in another PR. For instance, in `deprecated_references_for` package param seems to be both a String and Package. It's a little weird and I really think necessitates further investigation.

**Comment by cindygshopify** [comment]
I've pushed one more commit that moves the detect_stale_violations function that was in cli.rb into it's own class so we can get some separation of concerns in the Packwerk codebase. This required some refactoring of cli.rb as well by putting some common functions into a module. I've also updated the code to reflect some of the above mentioned comments, so please take a look. There are some sorbet signatures I have not added as it negatively affects the tests that I've written. The solution to that requires even more refactoring in my opinion. In the interest of time, I'd prefer not to drag out this PR more if it is related to further refactoring. I think this is better served by creating new GitHub issues around that so we can avoid scope creep.

File: lib/packwerk/cache_deprecated_references.rb
Comment (cindygshopify) [reply lib/packwerk/cache_deprecated_references.rb:15]: done

**Comment by tomstuart** [comment]
Thank you Cindy, the code looks good to me. My laptop is (very slowly) tophatting it now, then I’ll approve 👍

**Comment by tomstuart** [comment]
`packwerk update` works fine for this branch, so that looks good.

I’m sorry but I couldn’t successfully tophat `dev packages detect-stale-violations` in Shopify/shopify. When I tried to run it I got the error:

```
lib/packwerk/cli.rb:192:in `detect_stale_violations': uninitialized constant Packwerk::Cli::DetectStaleViolationsCommand (NameError)
```

I can fix that by putting `require "packwerk/commands/detect_stale_violations_command"` at the top of `lib/packwerk/cli.rb`, but then I get a different error:

```
lib/packwerk/commands/detect_stale_violations_command.rb:33:in `block (2 levels) in run': private method `file_processor' called for #<Packwerk::RunContext> (NoMethodError)
```

I could fix _that_ by making `RunContext#file_processor` public, but rather than continue chasing these I thought I’d report back and let you know it wasn’t working right. Was it working fine before? What’s changed?

**Comment by cindygshopify** [comment]
> Was it working fine before? What’s changed?

It was. I tested before all the refactoring shenanigans so that might've broken something. I figured it would have been covered by the test cases I wrote, but it seems like it wasn't. I'll take another look

**Comment by cindygshopify** [comment]
There was a minor change in another file that broke the functionality of detect-stale-violations, I just pushed with an amended change. Tophatting should work now.

**Comment by wildmaples** [comment]
> There was a minor change in another file that broke the functionality of detect-stale-violations, I just pushed with an amended change. Tophatting should work now.

What was the issue and how did you solve it?

File: lib/packwerk/cli.rb
Comment (wildmaples) [reply lib/packwerk/cli.rb]: https://github.com/Shopify/packwerk/pull/61/files/fe49aee8f6b023360c806093f6b05dc6e2ddd6b4#diff-ac5ae86965a90253f8d378798aa32f4041ec009d6b79c4420e88be6e67108e88R133 - I think this should work.

Within `#update_deprecations` method a little above in this file, it goes on to define a variable called `updating_deprecated_references`, which is a `reference_lister`. 

Reference lister is the same type as the class you are creating in this line, so the suggestion was to keep things consistent and readable in the other method by changing the variable name to reference_lister too. (It's a suggestion though, feel free to ignore!)

**Comment by cindygshopify** [comment]
> > There was a minor change in another file that broke the functionality of detect-stale-violations, I just pushed with an amended change. Tophatting should work now.
> 
> What was the issue and how did you solve it?

RunContext#file_processor was a public function that was being used in DetectStaleViolationsCommand. However due to a refactor outside of this PR, it got changed to be a private function. However since this function was previously being used outside of it's own class, another function was written to replace the functionality of the function calls outside its' class. Thus I replaced RunContext#file_processor with RunContext#process_file

File: lib/packwerk/cli.rb
Comment (wildmaples) [reply lib/packwerk/cli.rb:191]: Can you add an issue for this refactor some of the CLI logic out of the CLI so we don't forget?

Comment (wildmaples) [reply lib/packwerk/cli.rb:191]: Never mind, I just did it! https://github.com/Shopify/packwerk/issues/70

**Review by wildmaples — APPROVED** [review]
🎩 Tophatted and it works!!! 🎉 

This is a huge (and much needed) change, but I would prefer to include it in release `v1.0.3` instead of `v1.0.2`. 

Reason is because we are currently fixing a change for `v1.0.2` and if this change is merged and for some unexpected reason breaks, we'd have two back-to-back broken versions.

**Comment by cindygshopify** [comment]
> Reason is because we are currently fixing a change for `v1.0.2` and if this change is merged and for some unexpected reason breaks, we'd have two back-to-back broken versions.

Okay. Will we need to hold off merging this for a while then or would that not matter and just be something to do noted by the release manager when doing the releases?

File: lib/packwerk/cli.rb
Comment (doodzik) [inline lib/packwerk/cli.rb:198]: 🙌

**Review by doodzik — APPROVED** [review]
LGTM

**Comment by exterm** [comment]
Is anyone following up by using this to fix the stale violations check that is already in core CI?

**Comment by cindygshopify** [comment]
@exterm I'm going that

**Pull request merged**
